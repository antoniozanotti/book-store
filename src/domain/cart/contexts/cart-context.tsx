"use client";

import {
  createContext,
  useContext,
  useReducer,
  ReactNode,
  useEffect,
} from "react";
import { CartType } from "../types/cart-type";
import { CartContextType } from "../types/cart-context-type";
import { useLocalStorage } from "usehooks-ts";
import { useProductsByIdsQuery } from "@/domain/product/queries/use-products-by-ids-query";
import { cartReducer } from "../reducers/cart-reducer";

const initialState: CartType = { items: [] };

const CartContext = createContext<CartContextType>({
  state: initialState,
  dispatch: () => {},
  getCartItemsQty: () => {},
  cartItemsSubtotal: () => {},
});

export function CartProvider({ children }: { children: ReactNode }) {
  const [localState, setLocalState] = useLocalStorage<CartType>(
    "cart",
    initialState
  );
  const [state, dispatch] = useReducer(cartReducer, localState);
  const cartItems = state.items;
  const responses = useProductsByIdsQuery(cartItems.map((item) => item.id));

  function getCartItemsQty(): number {
    const cartItems = state.items;
    return cartItems.reduce((acc, cur) => {
      return acc + cur.quantity;
    }, 0);
  }

  function cartItemsSubtotal(): number {
    let subTotal = 0;
    cartItems.forEach((item) => {
      responses.some((response) => {
        const product = response.data;
        if (!product) {
          return true;
        }
        if (item.id == product.id) {
          subTotal += parseInt(product.price.toString()) * item.quantity;
          return true;
        }
      });
    });
    return subTotal;
  }

  useEffect(() => {
    setLocalState(state);
  }, [state, setLocalState]);

  return (
    <CartContext.Provider
      value={{ state, dispatch, getCartItemsQty, cartItemsSubtotal }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCartContext(): CartContextType {
  return useContext(CartContext);
}
