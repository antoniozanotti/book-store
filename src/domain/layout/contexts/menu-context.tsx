"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { MenuType } from "../types/menu-type";
import { MenuContextType } from "../types/menu-context-type";
import { menuReducer } from "../reducers/menu-reducer";

const initialState: MenuType = { isOpen: false };

const MenuContext = createContext<MenuContextType>({
  state: initialState,
  dispatch: () => {},
});

export function MenuProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(menuReducer, initialState);

  return (
    <MenuContext.Provider value={{ state, dispatch }}>
      {children}
    </MenuContext.Provider>
  );
}

export function useMenuContext(): MenuContextType {
  return useContext(MenuContext);
}
