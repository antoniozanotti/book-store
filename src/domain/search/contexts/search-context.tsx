"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { SearchContextType } from "../types/search-context-type";
import { useSearchParams } from "next/navigation";
import { getSearchParams } from "../services/search-service";
import { searchReducer } from "../reducers/search-reducer";

const SearchContext = createContext<SearchContextType>({} as SearchContextType);

export function SearchProvider({ children }: { children: ReactNode }) {
  const searchParams = useSearchParams();
  const { keyword, categories, minPrice, maxPrice, minYear, maxYear, orderBy } =
    getSearchParams(searchParams);
  const initialStateBySearchParams = {
    keyword: keyword,
    categories: categories,
    minprice: minPrice,
    maxprice: maxPrice,
    minyear: minYear,
    maxyear: maxYear,
    orderby: orderBy,
  };
  const [state, dispatch] = useReducer(
    searchReducer,
    initialStateBySearchParams
  );

  return (
    <SearchContext.Provider value={{ state, dispatch }}>
      {children}
    </SearchContext.Provider>
  );
}

export function useSearchContext(): SearchContextType {
  return useContext(SearchContext);
}
