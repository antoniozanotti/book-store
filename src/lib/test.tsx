import React, { ReactElement, ReactNode } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <div data-testid="container">{children}</div>;
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";
import { usePathname } from "next/navigation";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
    notFound: () => null,
    usePathname: () => null,
    useSearchParams: () => null,
  };
});
