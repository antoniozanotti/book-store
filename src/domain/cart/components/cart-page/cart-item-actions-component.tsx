"use client";

import { UiPriceComponent } from "@/domain/ui/components/ui-price-component";
import { UiButtonComponent } from "@/domain/ui/components/ui-button-component";
import { MinusIcon, PlusIcon } from "lucide-react";
import { useCartContext } from "../../contexts/cart-context";
import { CartActionEnum } from "../../enums/cart-action-enum";
import { ProductWithPageAndBookPrisma } from "@/domain/product/types/product-prisma";

export function CartItemActions({
  product,
  cartItemQty,
}: {
  product: ProductWithPageAndBookPrisma;
  cartItemQty: number;
}) {
  const cartDispatch = useCartContext().dispatch;
  return (
    <div className="flex gap-2 h-fit self-end col-span-2 lg:col-auto">
      <UiButtonComponent
        className="text-sm p-0 bg-transparent text-red-700"
        onClick={() =>
          cartDispatch({
            type: CartActionEnum.REMOVE_FROM_CART,
            payload: { id: product.id },
          })
        }
      >
        remove
      </UiButtonComponent>
      <div className="border-y border-primary-900 flex gap-2 items-center">
        <UiButtonComponent
          className="text-sm p-1.5"
          aria-label="decrease quantity"
          onClick={() =>
            cartDispatch({
              type: CartActionEnum.DECREASE_CART_ITEM_QUANTITY,
              payload: { id: product.id },
            })
          }
        >
          <MinusIcon />
        </UiButtonComponent>
        <div className="w-10 text-center text-sm" aria-label="quantity">
          {cartItemQty}
        </div>
        <UiButtonComponent
          className="text-sm p-1.5"
          aria-label="increase quantity"
          onClick={() =>
            cartDispatch({
              type: CartActionEnum.INCREASE_CART_ITEM_QUANTITY,
              payload: { id: product.id },
            })
          }
        >
          <PlusIcon />
        </UiButtonComponent>
      </div>
      <span className="font-bold grow flex items-end justify-end">
        <span aria-label="cart item subtotal">
          <UiPriceComponent
            value={cartItemQty * parseInt(product.price.toString())}
          />
        </span>
      </span>
    </div>
  );
}
