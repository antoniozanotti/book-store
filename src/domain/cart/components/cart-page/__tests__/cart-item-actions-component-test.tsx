import { render } from "@/lib/test";
import { CartItemActions } from "../cart-item-actions-component";
import { CartProvider } from "@/domain/cart/contexts/cart-context";
import { productWithPagEandBookMock } from "@/domain/product/__tests__/product-with-page-and-book-mock";
import { TanstackProvider } from "@/lib/tanstack-provider";

describe("CartItemActions", () => {
  it("should render", async () => {
    const { container } = render(
      <TanstackProvider>
        <CartProvider>
          <CartItemActions
            product={productWithPagEandBookMock}
            cartItemQty={1}
          />
        </CartProvider>
      </TanstackProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
