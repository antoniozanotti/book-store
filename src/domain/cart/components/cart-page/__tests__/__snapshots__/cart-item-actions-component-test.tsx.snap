// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemActions should render 1`] = `
<div>
  <div
    data-testid="container"
  >
    <div
      class="flex gap-2 h-fit self-end col-span-2 lg:col-auto"
    >
      <button
        class="bg-primary-900 font-bold p-4 text-primary-100 text-sm p-0 bg-transparent text-red-700"
      >
        remove
      </button>
      <div
        class="border-y border-primary-900 flex gap-2 items-center"
      >
        <button
          aria-label="decrease quantity"
          class="bg-primary-900 font-bold p-4 text-primary-100 text-sm p-1.5"
        >
          <svg
            class="lucide lucide-minus "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
          </svg>
        </button>
        <div
          aria-label="quantity"
          class="w-10 text-center text-sm"
        >
          1
        </div>
        <button
          aria-label="increase quantity"
          class="bg-primary-900 font-bold p-4 text-primary-100 text-sm p-1.5"
        >
          <svg
            class="lucide lucide-plus "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
        </button>
      </div>
      <span
        class="font-bold grow flex items-end justify-end"
      >
        <span
          aria-label="cart item subtotal"
        >
          $302.00
        </span>
      </span>
    </div>
  </div>
</div>
`;
